kore/src/Kore/Strategies/Goal.hs

```diff
@@ -40,94 +102,195 @@ See also: 'Strategy.pickFinal', 'extractUnproven'
+non-deterministic definition
 
-    -- | Apply 'Rule's to the goal in parallel.
-    derivePar
-        :: MonadSimplify m
-        => [Rule goal]
-        -> goal
-        -> Strategy.TransitionT (Rule goal) m (ProofState goal)
+> module ABC
+>   import DOMAINS
+>   syntax S ::= "a" | "b" | "c"
+>   rule [ab]: a => b
+>   rule [ac]: a => c
+> endmodule
 
-    -- | Apply 'Rule's to the goal in sequence.
-    deriveSeq
-        :: MonadSimplify m
-        => [Rule goal]
+this claim would be provable,
+
+> rule a => b [claim]
+
+but this claim would **not** be provable,
+
+> rule a => c [claim]
+
+because the algorithm would first apply semantic rule [ab], which prevents rule
+[ac] from being used.
+
+We decided to assume that the definition is deterministic because one-path
+verification is mainly used only for deterministic semantics and the assumption
+simplifies the implementation. However, this assumption is not an essential
+feature of the algorithm. You should not rely on this assumption elsewhere. This
+decision is subject to change without notice.
+
+This instance contains the default implementation for a one-path strategy. You can apply it to the
+first two arguments and pass the resulting function to 'Kore.Strategies.Verification.verify'.
+
+Things to note when implementing your own:
+
+1. The first step does not use the reachability claims
+
+2. You can return an infinite list.
+-}
+
+instance (SimplifierVariable variable) => Goal (OnePathRule variable) where
+
+    newtype Rule (OnePathRule variable) =
+        OnePathRewriteRule { unRule :: RewriteRule variable }
+        deriving (GHC.Generic, Show, Unparse)
+
+    type Prim (OnePathRule variable) =
+        ProofState.Prim (Rule (OnePathRule variable))
+
+    type ProofState (OnePathRule variable) a =
+        ProofState.ProofState a
+
+    transitionRule =
+        transitionRuleTemplate
+            TransitionRuleTemplate
+                { simplifyTemplate =
+                    simplify
+                , removeDestinationTemplate =
+                    removeDestination
+                , isTriviallyValidTemplate =
+                    isTriviallyValid
+                , deriveParTemplate =
+                    derivePar
+                , deriveSeqTemplate =
+                    deriveSeq
+                }
+
+    strategy goals rules =
+        onePathFirstStep rewrites
+        : repeat
+            ( onePathFollowupStep
+                coinductiveRewrites
+                rewrites
+            )
+      where
+        rewrites = rules
+        coinductiveRewrites =
+            OnePathRewriteRule
+            . RewriteRule
+            . getOnePathRule
+            <$> goals
+
+instance SOP.Generic (Rule (OnePathRule variable))
+
+instance SOP.HasDatatypeInfo (Rule (OnePathRule variable))
+
+instance Debug variable => Debug (Rule (OnePathRule variable))
+
+data TransitionRuleTemplate monad goal =
```

Comment:
I made this change here: https://github.com/kframework/kore/pull/955/commits/8d74980195f3ed5f99da7adba989c1c1b205da54

It looks cleaner to me this way, but if @ttuegel has any objections I'll revert this.

Reply:
> This record, and the way it is used seems to me like it could be replaced by a class.

I'm confused: It was a class, and un-making that class was the motivation for this pull request, because that class wasn't reusable.

Reply:
It's a class which is reusable for at least OnePath and AllPath

The alternative was a dictionary of functions which was also only used for OnePath and AllPath.